src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -3827,62 +3821,172 @@ VOID MethodTableBuilder::AllocateFieldDescs()
     }
 }
 
-//*******************************************************************************
-BOOL MethodTableBuilder::IsSelfReferencingStaticValueTypeField(mdToken     dwByValueClassToken,
-                                                               bmtInternalInfo* bmtInternal,
-                                                               const bmtGenericsInfo *bmtGenerics,
-                                                               PCCOR_SIGNATURE pMemberSignature,
-                                                               DWORD       cMemberSignature)
+CorElementType MethodTableBuilder::GetCorElementTypeOfTypeDefOrRefForStaticField(Module* module, mdToken typeDefOrRef)
 {
     STANDARD_VM_CONTRACT;
 
-    if (dwByValueClassToken != this->GetCl())
+    MethodTable *pMTFound = NULL;
+
+    if (TypeFromToken(typeDefOrRef) == mdtTypeDef)
     {
-        return FALSE;
+        pMTFound = module->LookupTypeDef(typeDefOrRef).AsMethodTable();
+    }
+    else if (TypeFromToken(typeDefOrRef) == mdtTypeRef)
+    {
+        pMTFound = module->LookupTypeRef(typeDefOrRef).AsMethodTable();
     }
 
-    if (!bmtGenerics->HasInstantiation())
+    if ((pMTFound == NULL) && bmtInternal->pType != NULL && (typeDefOrRef != GetCl()))
     {
-        return TRUE;
+        EX_TRY
+        {
+            pMTFound = ClassLoader::LoadTypeDefOrRefThrowing(module, typeDefOrRef, ClassLoader::ThrowIfNotFound, ClassLoader::PermitUninstDefOrRef, 0, CLASS_LOAD_APPROXPARENTS).AsMethodTable();
+        }
+        EX_CATCH
+        {
+            // If this failed, we can fall back to directly working with the type definition, so catch all the exceptions
+        }
+        EX_END_CATCH
     }
 
-    // The value class is generic.  Check that the signature of the field
-    // is _exactly_ equivalent to VC<!0, !1, !2, ...>.  Do this by consing up a fake
-    // signature.
-    DWORD nGenericArgs = bmtGenerics->GetNumGenericArgs();
-    CONSISTENCY_CHECK(nGenericArgs != 0);
+    if (pMTFound != NULL)
+    {
+        if (pMTFound->IsByRefLike())
+        {
+            BuildMethodTableThrowException(IDS_CLASSLOAD_BYREF_OR_BYREFLIKE_STATICFIELD);
+        }
 
-    SigBuilder sigBuilder;
+        if (pMTFound->IsEnum())
+        {
+            return pMTFound->GetApproxFieldDescListRaw()[0].GetFieldType();
+        }
+        else
+        {
+            _ASSERTE(pMTFound->IsValueType());
```

Comment:
What do we do to ensure that this will be true? Isn't it possible that we could have a static field of `object` here? The comments I found for `IsByRefLike` suggest that it's only true for TypedReference, Span, etc.

If the comment I read is wrong it then suggests that the IsByRefLike check above is banning reftype static fields, so I assume I'm missing something here.

Reply:
We only call GetCorElementTypeOfTypeDefOrRefForStaticField if the CorElementType in the static field signature is ELEMENT_TYPE_VALUETYPE. This obviously is vulnerable to malicious dlls which do not specify the correct element type, but CoreCLR per policy does not seek to validate the correctness of the assemblies that it inputs.